src/libraries/Common/src/System/Security/Cryptography/Oids.Shared.cs

```diff
@@ -65,5 +68,66 @@ private static Oid InitializeOid(string oidValue)
             return oid;
         }
 
+        internal static Oid GetSharedOrNewOid(ref AsnValueReader asnValueReader)
+        {
+            Oid? ret = GetSharedOrNullOid(ref asnValueReader);
+
+            if (ret is not null)
+            {
+                return ret;
+            }
+
+            string oidValue = asnValueReader.ReadObjectIdentifier();
+            return new Oid(oidValue, null);
+        }
+
+        internal static Oid? GetSharedOrNullOid(ref AsnValueReader asnValueReader)
+        {
+            Asn1Tag tag = asnValueReader.PeekTag();
+
+            // Any of these cases are going to be an invalid OID.  We'll let that get thrown naturally, the same as if
+            // it wasn't a match.
+            if (!tag.IsConstructed &&
+                (tag.TagClass != TagClass.Universal || tag.TagValue == (int)UniversalTagNumber.ObjectIdentifier))
+            {
+                ReadOnlySpan<byte> contentBytes = asnValueReader.PeekContentBytes();
+                Oid? ret = null;
+
+                switch (contentBytes.Length)
+                {
+                    case 3:
```

Comment:
I feel like this is going to be a really good candidate for list patterns, rather than manually writing the switch logic.

https://sharplab.io/#v2:EYLgtghglgdgPgAQEwEYCwAoBBmABM3AYVwG9NcL88EUAGXAWQApgBPAFwFMBtAXV1gAHAK7sAlKXKVpCAOwCYI9rgDOAdyjsAxgAtJGaYcrdaADwCs5gDS4ztACw272fgF4AfLgBESAHTnfe19sLyspI2kTC2tbUwcnOPs3Tx9/QMDQ8IiKKMsE+NjaADZk7z8AoKLMgwiAXwBuLNrMWqA=

Reply:
Yes, boo, hiss, new C# features. But the list pattern is way more readable :-).

Reply:
I knew (believed?) that the new switches did something like that for strings.  Didn't know it had a version that worked on spans.  (Changing from `byte[]` to `ReadOnlySpan<byte>` still compiles).

I'll reluctantly agree that it's a new feature that made the code nicer :smile:

Reply:
Also, I take it you don't object to the premise.  (The ability to use ReferenceEquals is going to be limited to System.Security.Cryptography.dll, but it should still save on bit banging, and a string allocation, for every OID we read that is well-known (and in this lookup)).

Reply:
I'm... not sure now that you draw attention to it. The creation of these Oids can race when they are lazily inited, and we can not use this pattern in S.S.C.Pkcs. It also feels fragile in that any refactoring or if a `new Oid(...)` ends up involved anywhere, it won't work.

I'm not "against" it either really, and I assume things will be well tested.

I'd sooner implement some kind of equality on Oid that does

```C#
if (ReferenceEquals(this, other)) return true;
return this.Value == other.Value;
```

It doesn't have to be public, either. Ideally it would be, but that starts the whole immutable OID conversation in my head again, and then I get sad.

Reply:
Yeah, the non-atomic lazy was gnawing at me last night, so I made an `internal` ValueEquals extension method this morning; seemed better than converting everything to locked loading.